### PR TITLE
[#777] Allow runtime parametrization of Azure blobstore configuration

### PR DESCRIPTION
**Pull request by fernandor777**
This PR fixes runtime parametrization on Azure blobstore configuration, leveraging placeholders resolving on blobstore creation runtime instead of Xstream resolvers.

Issue:
https://github.com/GeoWebCache/geowebcache/issues/777

### CODE REVIEW

**Comment by aaime** [comment]
QA checks failed:

```
[WARNING] /home/travis/build/GeoWebCache/geowebcache/geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java:[49,36] Integer(java.lang.String) in java.lang.Integer has been deprecated
[WARNING] /home/travis/build/GeoWebCache/geowebcache/geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java:[59,33] Boolean(java.lang.String) in java.lang.Boolean has been deprecated
```

In both cases use the valueOf static method instead, these constructors are bad for performance and are about to be removed in newer versions of Java.

To run these checks locally, add "-Dqa" to the maven call (warning, slow, best do it only in the module that got modified). This is valid for gwc but also gt and gs.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java [hunk 1]
```diff
@@ -60,36 +41,19 @@ public Object fromString(String str) {
...
-        return str;
-    }
-
     @Override
     public XStream getConfiguredXStream(XStream xs) {
         Class<AzureBlobStoreInfo> clazz = AzureBlobStoreInfo.class;
         xs.alias("AzureBlobStore", clazz);
-        xs.registerLocalConverter(clazz, "maxConnections", EnvironmentNullableIntConverter);
-        xs.registerLocalConverter(clazz, "proxyPort", EnvironmentNullableIntConverter);
+        xs.registerLocalConverter(clazz, "maxConnections", EnvironmentStringConverter);
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java]: No need to register custom converters any longer, the default behavior should do (please check).

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java [hunk 2]
```diff
@@ -18,37 +18,18 @@
...
-                public Object fromString(String str) {
-                    str = resolveFromEnv(str);
-                    if (Strings.isNullOrEmpty(str)) {
-                        return null;
-                    }
-                    return super.fromString(str);
-                }
-            };
-
     private static SingleValueConverter EnvironmentNullableBooleanConverter =
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java]: Likely not needed any longer, just check the default behavior in the case of "empty or null"

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java [hunk 3]
```diff
@@ -41,7 +41,7 @@
      */
     public static final int DEFAULT_CONNECTIONS = 100;
 
-    static Log log = LogFactory.getLog(AzureBlobStoreInfo.class);
+    private static final Logger LOGGER = Logger.getLogger(AzureBlobStoreInfo.class.getName());
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java]: GWC uniformly uses commons logging, please revert this change.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java [hunk 4]
```diff
@@ -123,12 +123,14 @@ public void setPrefix(String prefix) {
     }
 
     /** @return The maximum number of allowed open HTTP connections. */
-    public Integer getMaxConnections() {
-        return maxConnections == null ? DEFAULT_CONNECTIONS : maxConnections;
+    public String getMaxConnections() {
+        return maxConnections == null
+                ? Integer.valueOf(DEFAULT_CONNECTIONS).toString()
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java]: ``String.valueOf(DEFAULT_CONNECTIONS)`` would be more straightforward.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import org.geowebcache.GeoWebCacheTypedEnvironment;
+
+/**
+ * Azure Blobstore type-resolved data from a {@link AzureBlobStoreInfo} using enviroment variables
+ * if enabled.
+ */
+public class AzureBlobStoreData {
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java]: Type safe bean, yep, I would have likely done the same. Any actual need to make this class public, though?

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    public Integer getMaxConnections() {
+        return maxConnections;
+    }
+
+    public void setMaxConnections(Integer maxConnections) {
+        this.maxConnections = maxConnections;
+    }
+
+    public Boolean isUseHTTPS() {
```
Comment (aaime) [inline geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java]: Should allow parametrization of this one too --> String.

File: geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreConfigProviderTest.java [hunk 7]
```diff
@@ -18,35 +18,10 @@
...
-import org.geowebcache.GeoWebCacheExtensions;
-import org.junit.After;
-import org.junit.Before;
 import org.junit.Test;
-import org.springframework.context.support.ClassPathXmlApplicationContext;
 
 public class AzureBlobStoreConfigProviderTest {
 
-    @Before
-    public void setUp() throws Exception {
```
Comment (aaime) [inline geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreConfigProviderTest.java]: Please do not remove, the test is now to check that the environment variables can survive a xstream load (this is actually at the very core of the issue report), while there should be another test checking that variable expansion works.

File: geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreDataTest.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        GeoWebCacheExtensions gse = new GeoWebCacheExtensions();
+        gse.setApplicationContext(context);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        System.clearProperty("CONTAINER");
+        System.clearProperty("CONNECTIONS");
+        System.clearProperty("ACCOUNT_NAME");
+        System.clearProperty("ALLOW_ENV_PARAMETRIZATION");
```
Comment (aaime) [inline geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreDataTest.java]: Should clear ACCOUNT_KEY too

File: geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @author Fernando Mino, GeoSolutions S.A.S., Copyright 2019
+ */
+package org.geowebcache;
+
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+
+/** Enviroment variables placeholder resolver with declared types. */
+public class GeoWebCacheTypedEnvironment {
```
Comment (aaime) [inline geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java]: I don't see a need for a new class, an extra static method in GeoWebCacheEnvironment should do the trick.

File: geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * Resolve placeholders for a String parameter if enabled on configurations.
+     *
+     * @param value String value with optional placeholders.
+     * @param type Data Type to return
+     * @return Optional resolved value.
+     */
+    public <T extends Object> Optional<T> resolve(final String value, Class<T> type) {
```
Comment (aaime) [inline geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java]: I'd keep method name conventions intact, call it ``resolveValue`` like in ``GeoWebCacheEnvironment`` and it has the extra parameter for type conversion.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java (hunk 1 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreConfigProvider.java]: Indeed not needed anymore, removed.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java (hunk 3 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java]: Thanks, reverted.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java (hunk 4 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreInfo.java]: Fixed.

File: geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java (hunk 5 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/main/java/org/geowebcache/azure/AzureBlobStoreData.java]: Right, not public makes sense.

File: geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreDataTest.java (hunk 8 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreDataTest.java]: Added.

File: geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java (hunk 9 above)
Comment (fernandor777) [reply geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java]: Indeed, merged new method on GeoWebCacheEnvironment class.

File: geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java (hunk 10 above)
Comment (fernandor777) [reply geowebcache/core/src/main/java/org/geowebcache/GeoWebCacheTypedEnvironment.java]: Added as resolveValueIfEnabled since this method does the check for parameter-replacement flag enabled inside (the former 'resolveValue' method doesn't).

File: geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreConfigProviderTest.java (hunk 7 above)
Comment (fernandor777) [reply geowebcache/azureblob/src/test/java/org/geowebcache/azure/AzureBlobStoreConfigProviderTest.java]: Reverted.

**Comment by aaime** [comment]
Looks good, merged and hopefully kicked an automatic backport.
